### PR TITLE
Added IPNS publish feature

### PR DESCRIPTION
**Pull request by TalhaAnis**
This pull request provides a feature to publish new hash to IPNS every time there is a change in the hash of the directory. IPNS is slow procedure so if a hash is already being published to IPNS and there is a new hash available, the old IPNS publish process will be killed and a new IPNS publish process will start.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ import (
 var (
 	verbose = flag.Bool("verbose", false, "Print logs to stderr")
 	nodeURL = flag.String("node-addr", "/ip4/127.0.0.1/tcp/5001/", "IPFS node URL")
+	publish = flag.Bool("publish", false, "Option to publish to IPNS")
+	key     = flag.String("key", "", "key to publish to publish to IPNS")
```
Comment (crackcomm) [inline main.go]: could it be one flag?

File: shell/shell.go [hunk 2]
```diff
@@ -57,3 +58,37 @@ func RmLink(last, path string) (string, error) {
...
+// Publish - Publishes the ipfs hash to the provided ipns key
+func Publish(cmd *exec.Cmd, hash, key string) (*exec.Cmd, error) {
+	if !keyExists(key) {
+		return nil, errors.New("Key does not exist")
+	}
+	if cmd != nil {
+		cmd.Process.Kill()
+	}
+
+	cmd = exec.Command("ipfs", "name", "publish", hash)
```
Comment (crackcomm) [inline shell/shell.go]: what is the reason you don't use `Exec("name", "publish", hash)`?

File: main.go [hunk 3]
```diff
@@ -78,11 +83,28 @@ func main() {
 		log.Fatalf("watch error: %v", err)
 	}
 
-	fmt.Println(snc.Hash())
+	var cmd *exec.Cmd
```
Comment (crackcomm) [inline main.go]: what is the reason to re-use this pointer?

File: main.go [hunk 4]
```diff
@@ -78,11 +83,28 @@ func main() {
...
+	var err error
+	if *publish && hash != "" {
+		cmd, err = shell.Publish(cmd, hash, *key)
+		if err != nil {
+			fmt.Printf("Publish error: %s\n", err.Error())
+			os.Exit(1)
+		} else if printLogs {
+			fmt.Printf("Publishing to key %s\n", *key)
+		}
+	} else if !*publish {
```
Comment (crackcomm) [inline main.go]: this function should not be invoked if that's true

File: shell/shell.go [hunk 5]
```diff
@@ -57,3 +58,37 @@ func RmLink(last, path string) (string, error) {
...
 	return Exec("object", "patch", "add-link", last, path, hash)
 }
+
+func keyExists(key string) bool {
+	keys, err := Exec("key", "list", "-l")
+	if err != nil {
+		return false
+	}
+	lines := strings.Split(keys, "\n")
+	for _, line := range lines {
```
Comment (crackcomm) [inline shell/shell.go]: can you do `range strings.Split(keys, "\n")` for better readability? same for `keyNames`

File: shell/shell.go [hunk 6]
```diff
@@ -57,3 +58,37 @@ func RmLink(last, path string) (string, error) {
...
+			if keyName != "" && keyName == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Publish - Publishes the ipfs hash to the provided ipns key
+func Publish(cmd *exec.Cmd, hash, key string) (*exec.Cmd, error) {
```
Comment (crackcomm) [inline shell/shell.go]: follow the pattern of other functions

**Review by crackcomm — CHANGES_REQUESTED** [review]
I think it could be one flag but `--publish` is not explicit enough, `--ipns-key` without a default should do the same job.

File: main.go (hunk 1 above)
Comment (TalhaAnis) [reply main.go]: Yeah sure. We can make it --publish=<IPNS-key>. And for self we can say --publish=""

File: README.md [hunk 7]
```diff
@@ -21,6 +21,20 @@ $ go get -u github.com/ipfn/ipfs-sync
 $ ipfs-sync --node-addr=multiaddr <directory>
 ```
 
+Publish to IPNS
+```console
+$ ipfs-sync --node-addr=multiaddr --publish=true --key=QmdZjqVMbATh1YVdbtZA9j9aVkGKUXjM4aCHAT6yGiPY9R <directory>
```
Comment (crackcomm) [inline README.md]: please limit to one line

File: shell/shell.go (hunk 2 above)
Comment (TalhaAnis) [reply shell/shell.go]: I need *exec.Cmd object to save in the main function. In order to track that if the old process is already running or not. Exec doesn't return *exec.Cmd object.

File: main.go (hunk 1 above)
Comment (crackcomm) [reply main.go]: `--ipns-publish-key` or `--ipns-key` and `--ipns-key=self` for self

it makes it easier for new users to understand

File: main.go (hunk 3 above)
Comment (TalhaAnis) [reply main.go]: The same logic as explained in above comment.
"I need *exec.Cmd object to save in the main function. In order to track that if the old process is already running or not. Exec doesn't return *exec.Cmd object."

File: main.go (hunk 1 above)
Comment (crackcomm) [reply main.go]: or `--publish-key`

File: main.go (hunk 3 above)
Comment (crackcomm) [reply main.go]: in that case please start a new goroutine and use a channel to send updates to this goroutine which publishes

File: shell/shell.go (hunk 2 above)
Comment (crackcomm) [reply shell/shell.go]: please use a goroutine and follow the API, using only single process

File: main.go (hunk 4 above)
Comment (TalhaAnis) [reply main.go]: checkPublish is function to keep main() clean and avoid repetition. This function checks if publish is true then publishes to IPNS else it just prints the hashes(Original functionality)

File: shell/shell.go (hunk 5 above)
Comment (TalhaAnis) [reply shell/shell.go]: Okay

File: main.go (hunk 4 above)
Comment (crackcomm) [reply main.go]: as I mentioned, please use `if !*publish { go publisher }` and then in loop use `if !*publish { chan <- hash }`

File: shell/shell.go (hunk 2 above)
Comment (TalhaAnis) [reply shell/shell.go]: Okay

File: main.go (hunk 3 above)
Comment (TalhaAnis) [reply main.go]: Okay

File: shell/shell.go (hunk 6 above)
Comment (crackcomm) [reply shell/shell.go]: On second thought you can extend this API with `context.Context` to use [`WithTimeout`](https://golang.org/pkg/context/#WithTimeout).

**Comment by crackcomm** [comment]
[WithCancel](https://golang.org/pkg/context/#WithCancel) can be used to cancel a process. [WithTimeout](https://golang.org/pkg/context/#WithTimeout) can be used to start publishing with a timeout.

Edit: [CommandContext](https://golang.org/pkg/os/exec/#CommandContext) can be used to create a command cancelable using context primitives. This command can have its own goroutine.

File: shell/shell.go (hunk 6 above)
Comment (TalhaAnis) [reply shell/shell.go]: I tried various method including context and sending channels. All of them failed. The only approach that worked was cmd.Process.Kill(). The issue is that exec.Command is a blocking command and that's why context and channel approach didn't work here

Comment (crackcomm) [reply shell/shell.go]: https://golang.org/pkg/os/exec/#CommandContext

Comment (TalhaAnis) [reply shell/shell.go]: Yes but this is only based on timeout. We need a manual trigger whenever there is a new hash we have to kill the old process.

Comment (crackcomm) [reply shell/shell.go]: In main you would create a channel that hashes are published to and start a goroutine that listens on this channel.
Hash received on this channel would execute a command in goroutine with context. Cancel function would be stored as local variable in goroutine. Whenever new hash is received `cancel()` is called.

Comment (crackcomm) [reply shell/shell.go]: In fact you don't need to store `cancel()` just `defer cancel()` in a loop with `select{}` or `<-chan`. I didn't use Go in a long time so you need to figure it out, I'm sure you will =)

Comment (TalhaAnis) [reply shell/shell.go]: Okay will work on these structural changes tomorrow.

File: main.go (hunk 1 above)
Comment (TalhaAnis) [reply main.go]: Done

File: shell/shell.go (hunk 5 above)
Comment (TalhaAnis) [reply shell/shell.go]: Done

File: README.md (hunk 7 above)
Comment (TalhaAnis) [reply README.md]: Done

File: shell/shell.go [hunk 8]
```diff
@@ -78,32 +80,36 @@ func keyExists(key string) bool {
...
-			select {
-			case hash := <-ch:
-				if key != "" && hash != "" {
-					cancel()
+			if hash := <-ch; hash != "" {
+				if key != "" {
+					if cancel != nil {
+						cancel()
+					}
 					ctx, cancel = context.WithCancel(context.Background())
```
Comment (crackcomm) [inline shell/shell.go]: Move `var ctx context.Context` up here.

**Comment by crackcomm** [comment]
Can you make this function accept string channel so it's possible to do:

```Go
	snc, err := sync.Watch(*nodeURL, path, shell.AddOptions{
		Ignore:          ignore,
		IgnoreRulesPath: *ignoreRulesPath,
	})
	if err != nil {
		log.Fatalf("watch error: %v", err)
	}

	fmt.Println(snc.Hash())

	go sync.Publish(key, snc.Events())
```

The function `shell.Publish` should be like so:

```Go
func Publish(key, hash string) (string, error) {
	return Exec(...)
}
```

Then the function `sync.Publish` should have interface like so:

```Go
func Publish(key string, hash <-chan string) error
```

**Comment by crackcomm** [comment]
Take in mind `Publish` should not print anything and is just an example.

File: shell/shell.go (hunk 8 above)
Comment (TalhaAnis) [reply shell/shell.go]: It has to be outside the loop

**Comment by crackcomm** [comment]
```Go
if sync.events != nil && sync.hash != hash {
  sync.events <- hash
}
if hash != "" {
  sync.hash = hash
}
```
Should be actually:
```Go
if sync.events != nil && sync.hash != hash && hash != "" {
  sync.events <- hash
}
if hash != "" {
  sync.hash = hash
}
```

It provides latest hash for accessor:

```Go
// Hash - Current hash.
func (sync *Synchronizer) Hash() string {
	return sync.hash
}
```

**Review by TalhaAnis — COMMENTED** [review]
> Can you make this function accept string channel so it's possible to do:
> 
> ```go
> 	snc, err := sync.Watch(*nodeURL, path, shell.AddOptions{
> 		Ignore:          ignore,
> 		IgnoreRulesPath: *ignoreRulesPath,
> 	})
> 	if err != nil {
> 		log.Fatalf("watch error: %v", err)
> 	}
> 
> 	fmt.Println(snc.Hash())
> 
> 	go sync.Publish(key, snc.Events())
> ```
> 
> The function `shell.Publish` should be like so:
> 
> ```go
> func Publish(key, hash string) (string, error) {
> 	return Exec(...)
> }
> ```
> 
> Then the function `sync.Publish` should have interface like so:
> 
> ```go
> func Publish(key string, hash <-chan string) error
> ```

Exec(...) doesn't have CommandContext
Plus we have to do checks if key == ""

**Comment by crackcomm** [comment]
It could be a great feature to use context for all commands and cancel `ipfs add` on file change. I see no problem with breaking the API and adding context to all methods.

**Comment by TalhaAnis** [comment]
Okay

**Comment by TalhaAnis** [comment]
This is a big change I believe. I have to allocate time for that. If possible in week else on next weekend. Submitting other changes. I think for now we can merge this and fix this later

**Comment by crackcomm** [comment]
I can pick it up where you left it and fix that if you're not comfortable with it.

**Comment by TalhaAnis** [comment]
I wanted to do but I have to go to sleep for next day at office.

**Comment by crackcomm** [comment]
It's okay but I won't merge to master tonight in that case.

**Comment by TalhaAnis** [comment]
Okay I'll try to implement in this week

**Comment by TalhaAnis** [comment]
@crackcomm This is the best I was able to achieve. I believe that Exec() function cannot be implemented with context because we have to initialize a context before each call to the function. Also I believe that the code is perfectly working fine, further changes could just be moving code from here to there. Further Up to you :)

**Comment by crackcomm** [comment]
It's merged in [c6e6a](https://github.com/ipfn/ipfs-sync/commit/c6e6a5d377378cb2fb908c89623cead326085b2b).

**Pull request closed without merging**